gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java

```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
 
     TypeAdapter chosen = delegate;
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {

```

Comment:
I see your point. I think we should fix the precedence for JsonAdapter as that is the expected behavior by clients. I think JsonAdapter is new enough to not have serious incompatibility issues (as we discovered with this primitive-related behavior).
I will update the pull request.